### PR TITLE
Use `Sampler` to determine whether to fill stack trace

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

`Flags.verboseExceptions()` gives a user only two choices - enabling
stacktrace for all exceptions or not. We could let a user take
trade-offs by allowing him or her to specify a `Sampler` specification.

Modifications:

- Accept a `Sampler` specification string as a value of the
  `com.linecorp.armeria.verboseExceptions` system property.
  - Use the default specification of `rate-limited=10`, which allows 10
    traces per second per exception type.
  - Note that `true` and `false` are translated into `always` and
    `never`.
- Add `ExceptionSampler`
- Replace `Flags.verboseExceptions()` with `verboseExceptionSampler()`
  and `verboseExceptionSamplerSpec()`
- The exceptions which never had stack traces can now have stack trace.
- Add a benchmark

Result:

- A user will get the stack trace for some exceptions out of the box,
  which is a better experience.
  - Previously, a user had to set `true` to the flag manually.
- A user can trade-off between the cost of filling in stack trace and
  the debuggability in a fine-grained manner.

Benchmark result:

```
Benchmark                                              Mode  Cnt          Score          Error  Units
ExceptionSamplerBenchmark.baseline_singleton          thrpt    5  115913078.321 ±  2470676.668  ops/s
ExceptionSamplerBenchmark.baseline_withStackTrace     thrpt    5     435184.216 ±    68777.567  ops/s
ExceptionSamplerBenchmark.baseline_withoutStackTrace  thrpt    5   51507410.813 ±  4643400.389  ops/s
ExceptionSamplerBenchmark.sampler_always              thrpt    5     460633.461 ±    23792.691  ops/s
ExceptionSamplerBenchmark.sampler_never               thrpt    5   65191830.351 ± 10603815.026  ops/s
ExceptionSamplerBenchmark.sampler_random_1            thrpt    5   22901381.973 ±   992182.987  ops/s
ExceptionSamplerBenchmark.sampler_random_10           thrpt    5    4131608.570 ±    62347.977  ops/s
ExceptionSamplerBenchmark.sampler_rateLimited_1       thrpt    5   16322277.831 ±   591082.048  ops/s
ExceptionSamplerBenchmark.sampler_rateLimited_10      thrpt    5   15205268.392 ±   339038.406  ops/s
```

### CODE REVIEW

**Comment by trustin** [comment]
@adriancole You might find this interesting.

**Comment by minwoox** [comment]
What is the rate that you used in the benchmark test?

File: benchmarks/src/jmh/java/com/linecorp/armeria/common/ExceptionSamplerBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+        ExceptionA() {}
+
+        private ExceptionA(@SuppressWarnings("unused") boolean dummy) {
+            super(null, null, false, false);
+        }
+    }
+
+    private static final class ExceptionB extends Throwable {
+        private static final long serialVersionUID = 7024319028414180839L;
```
Comment (minwoox) [inline benchmarks/src/jmh/java/com/linecorp/armeria/common/ExceptionSamplerBenchmark.java]: same `serialVersionUID` with the one from `ExceptionA`. 😉

File: core/src/main/java/com/linecorp/armeria/common/ExceptionSampler.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package com.linecorp.armeria.common;
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/ExceptionSampler.java]: nit: missing copyright

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 3]
```diff
@@ -854,4 +899,5 @@ private static String getLowerCased(String fullName) {
     }
 
     private Flags() {}
+
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java]: nit: revert?

File: core/src/main/java/com/linecorp/armeria/common/ExceptionSampler.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+final class ExceptionSampler implements Sampler<Class<? extends Throwable>> {
+
+    private final Map<Class<? extends Throwable>, Sampler<Class<? extends Throwable>>> samplers =
+            new NonBlockingHashMap<>();
+    private final Function<? super Class<? extends Throwable>,
+            ? extends Sampler<Class<? extends Throwable>>> samplerFactory;
+    private final String spec;
+
+    ExceptionSampler(String spec) {
+        samplerFactory = unused -> Sampler.of(spec);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/ExceptionSampler.java]: Is there any chance that a user wants to set different specs for different exceptions?

**Comment by trustin** [comment]
Updated benchmark result:

```
Benchmark                                              Mode  Cnt          Score         Error  Units
ExceptionSamplerBenchmark.baseline_singleton          thrpt    5  114632886.246 ± 3041939.347  ops/s
ExceptionSamplerBenchmark.baseline_withStackTrace     thrpt    5     405581.893 ±   83689.688  ops/s
ExceptionSamplerBenchmark.baseline_withoutStackTrace  thrpt    5   51689297.355 ± 3255826.929  ops/s
ExceptionSamplerBenchmark.sampler_always              thrpt    5     416875.330 ±   86907.919  ops/s
ExceptionSamplerBenchmark.sampler_never               thrpt    5   68404141.379 ± 7251312.879  ops/s
ExceptionSamplerBenchmark.sampler_random_1            thrpt    5   22103987.090 ± 2862974.467  ops/s
ExceptionSamplerBenchmark.sampler_random_10           thrpt    5    3853851.348 ±  551755.802  ops/s
ExceptionSamplerBenchmark.sampler_rateLimited_1       thrpt    5   15737672.948 ±  950162.974  ops/s
ExceptionSamplerBenchmark.sampler_rateLimited_10      thrpt    5   15963973.888 ±  841312.986  ops/s
```

File: core/src/main/java/com/linecorp/armeria/common/ExceptionSampler.java (hunk 4 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/ExceptionSampler.java]: I'm not sure. Perhaps we can consider that when someone asks for it? :wink:

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/ExceptionSampler.java]: Added a comment about this idea to `Flags`

File: core/src/main/java/com/linecorp/armeria/common/ExceptionSampler.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        this.spec = spec;
+    }
+
+    @Override
+    public boolean isSampled(Class<? extends Throwable> exceptionType) {
```
Comment (adriancole) [inline core/src/main/java/com/linecorp/armeria/common/ExceptionSampler.java]: so the rationale for the type parameter is solving the overhead problem caused by allocating an exception, correct?

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/ExceptionSampler.java]: Yeah, kind of. If we accepted a `Throwable`, we will see a chicken and egg problem. :smile:

**Comment by trustin** [comment]
`sampler_rateLimited_10` was tainted by wrong input. Here's the correct one:

```
Benchmark                                              Mode  Cnt          Score          Error  Units
ExceptionSamplerBenchmark.baseline_singleton          thrpt    5  115913078.321 ±  2470676.668  ops/s
ExceptionSamplerBenchmark.baseline_withStackTrace     thrpt    5     435184.216 ±    68777.567  ops/s
ExceptionSamplerBenchmark.baseline_withoutStackTrace  thrpt    5   51507410.813 ±  4643400.389  ops/s
ExceptionSamplerBenchmark.sampler_always              thrpt    5     460633.461 ±    23792.691  ops/s
ExceptionSamplerBenchmark.sampler_never               thrpt    5   65191830.351 ± 10603815.026  ops/s
ExceptionSamplerBenchmark.sampler_random_1            thrpt    5   22901381.973 ±   992182.987  ops/s
ExceptionSamplerBenchmark.sampler_random_10           thrpt    5    4131608.570 ±    62347.977  ops/s
ExceptionSamplerBenchmark.sampler_rateLimited_1       thrpt    5   16322277.831 ±   591082.048  ops/s
ExceptionSamplerBenchmark.sampler_rateLimited_10      thrpt    5   15205268.392 ±   339038.406  ops/s
```

File: grpc-protocol/src/main/java/com/linecorp/armeria/common/grpc/protocol/ArmeriaStatusException.java [hunk 6]
```diff
@@ -23,7 +23,7 @@
  */
 public class ArmeriaStatusException extends RuntimeException {
 
-    public static final long serialVersionUID = -8370257107063108923L;
+    private static final long serialVersionUID = -8370257107063108923L;
```
Comment (minwoox) [inline grpc-protocol/src/main/java/com/linecorp/armeria/common/grpc/protocol/ArmeriaStatusException.java]: 😱

**Comment by minwoox** [comment]
still LGTM 👍

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 7]
```diff
@@ -95,16 +95,19 @@
 
         switch (spec) {
             case "true":
+            case "always":
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java]: 👍

**Pull request merged**